bench: Prevent thread oversubscription and decreases the variance of result values

hebasto opened the pull request:
Split out from #18710.

Some results (borrowed from #18710):
![89121718-a3329800-d4c1-11ea-8bd1-66da20619696](https://user-images.githubusercontent.com/32963518/90146614-ecb89800-dd89-11ea-80fe-bac0e46e735e.png)

hebasto commented:
cc @martinus @JeremyRubin

JeremyRubin commented:
Hmmmm...

I'd maybe rather just abort the test and return early if only 1 thread gets made because we shouldn't ever be running with the checkqueue on a single core machine...

hebasto commented:
@JeremyRubin 
> Hmmmm...
> 
> I'd maybe rather just abort the test and return early if only 1 thread gets made because we shouldn't ever be running with the checkqueue on a single core machine...

We do not know for sure who many CPU cores/threads users dedicate to Bitcoin Core :)

With this PR benchmark results are quite consistent:
```
$ taskset --cpu-list 0-7 src/bench/bench_bitcoin -filter=CCheckQueueSpeedPrevectorJob
|              ns/job |               job/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|              177.62 |        5,629,902.65 |    1.9% |      0.01 | `CCheckQueueSpeedPrevectorJob`
```
```
$ taskset --cpu-list 0-3 src/bench/bench_bitcoin -filter=CCheckQueueSpeedPrevectorJob
|              ns/job |               job/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|              176.05 |        5,680,327.83 |    3.3% |      0.01 | `CCheckQueueSpeedPrevectorJob`
```
```
$ taskset --cpu-list 0-1 src/bench/bench_bitcoin -filter=CCheckQueueSpeedPrevectorJob
|              ns/job |               job/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|               55.45 |       18,034,367.56 |    2.6% |      0.00 | `CCheckQueueSpeedPrevectorJob`
```
```
$ taskset --cpu-list 0 src/bench/bench_bitcoin -filter=CCheckQueueSpeedPrevectorJob
|              ns/job |               job/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|               54.73 |       18,271,118.07 |    4.0% |      0.00 | `CCheckQueueSpeedPrevectorJob`
```

And early return do not disable the test, right?

laanwj commented:
If it's a benchmark that benchmarks thread synchronization (which is, I think, the only valid reason to run a benchmark multi-threaded) then I agree running it without threads makes little sense.

hebasto committed:
bench: Allow skip benchmark

Co-authored-by: Martin Ankerl <Martin.Ankerl@gmail.com>

hebasto committed:
bench: Prevent thread oversubscription

This change decreases the variance of benchmark results.

hebasto commented:
Updated d786765721e02e7d2c888c8276c02064e5056593 -> a65c1adc7638c2b6f2038cbe11a637b285dd8dd2 ([19710.01](https://github.com/hebasto/bitcoin/commits/19710.01) -> [19710.02](https://github.com/hebasto/bitcoin/commits/19710.02), [diff](https://github.com/hebasto/bitcoin/compare/19710.01..19710.02)).

Addressed:
- @JeremyRubin's [comment](https://github.com/bitcoin/bitcoin/pull/19710#issuecomment-673547482):
> Hmmmm...
> 
> I'd maybe rather just abort the test and return early if only 1 thread gets made because we shouldn't ever be running with the checkqueue on a single core machine...

- @laanwj's [comment](https://github.com/bitcoin/bitcoin/pull/19710#issuecomment-674016320)

> If it's a benchmark that benchmarks thread synchronization/interaction (which is, I think, the only valid reason to run a benchmark multi-threaded) then I agree running it without threads makes little sense.

MarcoFalke commented:
cr ACK a65c1adc7638c2b6f2038cbe11a637b285dd8dd2

vasild commented on src/bench/checkqueue.cpp:
```diff
@@ -26,6 +24,10 @@ static const unsigned int QUEUE_BATCH_SIZE = 128;
 // and there is a little bit of work done between calls to Add.
 static void CCheckQueueSpeedPrevectorJob(benchmark::Bench& bench)
 {
+    if (GetNumCores() < 2) {
+        return;
+    }
```
I think this warrants a comment explaining "why?".

Also, maybe don't drop the `MIN_CORES` constant and use it here instead of the magic number "2"?

vasild commented on src/bench/checkqueue.cpp:
```diff
@@ -44,7 +46,7 @@ static void CCheckQueueSpeedPrevectorJob(benchmark::Bench& bench)
     CCheckQueue<PrevectorJob> queue {QUEUE_BATCH_SIZE};
     boost::thread_group tg;
-    for (auto x = 0; x < std::max(MIN_CORES, GetNumCores()); ++x) {
+    for (auto x = 0; x < GetNumCores() - 1; ++x) {
```
Maybe also comment here "why we do - 1".

vasild reviewed (approved):
ACK a65c1adc

hebasto commented:
Updated a65c1adc7638c2b6f2038cbe11a637b285dd8dd2 -> 4f9164389a078f64258f3d7f2ecda5ab68968e8c ([pr19710.02](https://github.com/hebasto/bitcoin/commits/pr19710.02) -> [pr19710.03](https://github.com/hebasto/bitcoin/commits/pr19710.03), [diff](https://github.com/hebasto/bitcoin/compare/pr19710.02..pr19710.03)).

Addressed @vasild's comments:
- https://github.com/bitcoin/bitcoin/pull/19710#discussion_r476578354:
> Also, maybe don't drop the MIN_CORES constant and use it here instead of the magic number "2"?

- https://github.com/bitcoin/bitcoin/pull/19710#discussion_r476579234
> Maybe also comment here "why we do - 1".

promag commented on src/bench/checkqueue.cpp:
```diff
@@ -26,6 +26,10 @@ static const unsigned int QUEUE_BATCH_SIZE = 128;
 // and there is a little bit of work done between calls to Add.
 static void CCheckQueueSpeedPrevectorJob(benchmark::Bench& bench)
 {
+    if (GetNumCores() < MIN_CORES) {
```
4f9164389a078f64258f3d7f2ecda5ab68968e8c

nit, single line and also add comment.

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/19710#issuecomment-680261215).

promag reviewed (commented):
Concept ACK.

hebasto commented:
Updated 4f9164389a078f64258f3d7f2ecda5ab68968e8c -> 5e7acd65443d3fc7be084ea81be3089e4b28bc9a ([pr19710.03](https://github.com/hebasto/bitcoin/commits/pr19710.03) -> [pr19710.04](https://github.com/hebasto/bitcoin/commits/pr19710.04), [diff](https://github.com/hebasto/bitcoin/compare/pr19710.03..pr19710.04)).

Addressed @promag's [comment](https://github.com/bitcoin/bitcoin/pull/19710#discussion_r476722701):
> nit, single line and also add comment.

promag commented:
Code review ACK 5e7acd65443d3fc7be084ea81be3089e4b28bc9a.

vasild reviewed (approved):
ACK 5e7acd654

Thanks for the comments!

hebasto commented:
@MarcoFalke Mind having another look at this PR?

promag commented on src/bench/checkqueue.cpp:
```diff
@@ -26,6 +26,9 @@ static const unsigned int QUEUE_BATCH_SIZE = 128;
 static void CCheckQueueSpeedPrevectorJob(benchmark::Bench& bench)
 {
+    // We shouldn't ever be running with the checkqueue on a single core machine.
+    if (GetNumCores() < MIN_CORES) return;
```
Maybe ditch `MIN_CORES` and change this to `GetNumCores() == 1`?

hebasto replied:
https://en.cppreference.com/w/cpp/thread/thread/hardware_concurrency:
> If the value is not well defined or not computable, returns ​`0`​.

hebasto replied:
Maybe `<=1` ?

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/19710#issuecomment-680785136).

hebasto commented:
Updated 5e7acd65443d3fc7be084ea81be3089e4b28bc9a -> 3edc4e34fe2f92e7066c1455f5e42af2fdb43b99 ([pr19710.04](https://github.com/hebasto/bitcoin/commits/pr19710.04) -> [pr19710.05](https://github.com/hebasto/bitcoin/commits/pr19710.05), [diff](https://github.com/hebasto/bitcoin/compare/pr19710.04..pr19710.05)).

Addressed @promag's [comment](https://github.com/bitcoin/bitcoin/pull/19710#discussion_r477172550):
> Maybe ditch `MIN_CORES` and change this to `GetNumCores() == 1`?

fjahr commented:
Code review ACK 3edc4e34fe2f92e7066c1455f5e42af2fdb43b99

promag commented:
ACK
